proposals/0000-swiftpm-binary-dependencies.md

```diff
@@ -0,0 +1,435 @@
+   make available as open source.
+ * A large company has an internal team which wants to deliver a Swift package
+   for use in their iOS applications, but for for business reasons cannot publish
+   the source code.
+
+This proposal defines a new SwiftPM feature to allow SwiftPM to accept some
+forms of "binary packages". This proposal is intentionally written to be a
+address the above use cases *explicitly*, it **does not** define a general
+purpose "binary artifact" mechanism intended to address other use cases (such as
+accelerating build performance). The motivations for this are discussed in more
+detail below.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)
+
+Pitch Thread: https://forums.swift.org/t/spm-support-for-binaries-distribution/25549/24
+                        
+## Motivation
+
+SwiftPM has a large appeal to certain developer communities, like the iOS
+ecosystem, where it is currently very common to rely on closed source
+dependencies such as Firebase, GoogleAnalytics, Adjust and many more. Existing
+package managers like Cocoapods support these use cases. By adding such support
+to SwiftPM, we will unblock substantially more adoption of SwiftPM within those
+communities.
+
+Prior to Swift 5.1, the Swift compiler itself did not expose all of the features
+(like ABI compatibility) required to build a workable solution. Now that those
+features are present, it makes sense to reevaluate the role of binary packages.
+
+The goal of this proposal is to make *consumption* of binary packages as
+described above *easy*, *intuitive*, *safe*, and *consistent*. This proposal
+**does not** attempt to provide any affordances for the creation of the binary
+package itself. The overall intent of this proposal is to allow consumption of
+binary packages *where necessary*, but not to encourage their use or facilitate a
+transition from the existing source-based ecosystem to a binary one.
+
+This proposal is also focused at packages which come exclusively in binary form,
+it explicitly **does not** introduce a mechanism which allows a package to be
+present in either source or binary form. See alternatives considered for more
+information on this choice.
+
+## Proposed solution
+
+To enable binary dependencies we have to make changes in the `Package.swift` manifest file. First, we propose to add a new target type which describes a binary target. An example of such a package can be seen below:
+
+```swift
+let package = Package(
+    name: "SomePackage",
+    products: [
+        .library(name: "SomePackage", targets: ["SomePackageLib"])
+    ],
+    targets: [
+        .binaryTarget(
+            name: "SomePackageLib",
+            artifacts: [
+                .artifact(
+                    source: .url("https://github.com/some/package/releases/download/1.0.0/SomePackage-1.0.0.zip"),
+                    .when(llvmTriplets: [.init(archType: .x86_64, vendor: .apple, osType: .macOSX)])
```

Comment:
I think we can remove the artifact condition for now since we want to restrict ourselves to Apple platforms and XCFrameworks-only for the first version. The XCFramework will provide the platform conditionality we need.